src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs

```diff
@@ -196,6 +207,77 @@ static ImmutableArray<AssetsFileLogMessage> ParseLogMessages(LockFile lockFile,
             }
 
             return builder.ToImmutable();
+
+            Dictionary<string, LogLevel> BuildLevelByLibrary()
+            {
+                Dictionary<string, HashSet<string>> ancestorsByLibrary = new(StringComparer.OrdinalIgnoreCase);
+
+                // Build a map from child-to-ancestors
+                foreach (LockFileTargetLibrary lockFileLibrary in lockFileTarget.Libraries)
+                {
+                    string? parentId = lockFileLibrary.Name;
+
+                    if (parentId is null)
+                    {
+                        continue;
+                    }
+
+                    foreach (var dep in lockFileLibrary.Dependencies)
+                    {
+                        string childId = dep.Id;
+
+                        if (!ancestorsByLibrary.TryGetValue(childId, out HashSet<string>? ancestors))
+                        {
+                            ancestors = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                            ancestorsByLibrary.Add(childId, ancestors);
+                        }
+
+                        ancestors.Add(parentId);
+                    }
+                }
+
+                // Walk up the tree from each log message's library, propagating the highest level found.
+                Dictionary<string, LogLevel> levelByLibrary = new(StringComparer.OrdinalIgnoreCase);
+
+                foreach (AssetsFileLogMessage message in logMessages)
+                {
+                    Integrate(message.LibraryName, message.Level, visited: []);
+                }
+
+                return levelByLibrary;
+
+                void Integrate(string id, LogLevel level, HashSet<string> visited)
```

Comment:
I think it would be better if we make this non-recursive so that we don't run into a stack overflow problem,

Reply:
We highly likely won't have a tree that deep but better safe

Reply:
I considered this, but the graph would need to be 3000+ references deep to hit a stack overflow exception. Avoiding recursion requires heap allocation. I don't think it's a good tradeoff given the situation here. I'm already slightly concerned about the additional allocations needed here, especially for their impact on search in Solution Explorer which triggers all these snapshots to be created across the solution.